src/patterns/scheduled-lambda.ts

```diff
@@ -0,0 +1,27 @@
+import type { Schedule } from "@aws-cdk/aws-events";
+import type { GuStack } from "../constructs/core";
+import { GuLambdaFunction } from "../constructs/lambda";
+import type { GuFunctionProps } from "../constructs/lambda";
+import type { ErrorPercentageMonitoring } from "./lambda-alarms";
+import { GuLambdaErrorPercentageAlarm } from "./lambda-alarms";
+
+interface GuScheduledLambdaProps extends Omit<GuFunctionProps, "rules" | "apis"> {
+  schedule: Schedule;
+  monitoringConfiguration?: ErrorPercentageMonitoring;
```

Comment:
I wondered about making the opt-out for this more explicit. In Scala I would do something like:

```
sealed trait LambdaMonitoring
case object NoMonitoring extends LambdaMonitoring
case class ErrorPercentageMonitoring(toleratedErrorPercentage: Int, snsTopicArn: String, numberOfFiveMinutePeriodsToEvaluate: Option[Int]=Some(5)) extends LambdaMonitoring
```

The interface would then be:

```
monitoringConfiguration: LambdaMonitoring
```

So to opt-out, you'd have to state:

```monitoringConfiguration: NoMonitoring```

This would make it clearer (and highlight to reviewers) that you were intentionally omitting this (hopefully for a good reason e.g. you prefer some different alerting system to CloudWatch), rather than just being (another) optional property that can be easily forgotten.

Any opinions on whether we'd want to do this? If so, any suggestions on how we could go about this in TypeScript? I tried a couple of options using [Unions](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions) but they were all pretty ugly!

Reply:
Yeah, that would be nice. I don't know of a nice way of doing that in typescript although I often feel like there must be one that I'm missing.

Reply:
Does [this](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgDJwLYCMAmcCyA9iMGIVKAObIDeyAvgFCOiSyIoByhRJZFIahAAekEDgDOaTLgLFS5KrQbNW0eEmQBRKFHIAFaEnBxKEXgoFDREcVPTY8F-kpqNkyMgBtocSDh09KEMoYzBTCAAuZBAAV2xoABpkd2QJEAkAFUIAB2AEAEEoEGiJMCtk1LiEqAB5GAAxYAA3c1BYyBDgQklsrWa4L1i-CAB+aOqsaEYmRhhYkAQwbpBkHEIAZUIMCDAACyoACmFoh1lnRUFkAF5aemQ4KW4LqwBKZVmEYjLkE+Rn+QuK63OhML4ZMDIACe0UCBiMtnCZheShBnkIPigIwCunhoUREWiAEZkuksrl8kUSsgAEQwQiEGkqRjrLY7fZHYSvFmbba7A6CQ5Q15AA) work? (I'm not too certain on the `as` though!)

Reply:
Thanks for the suggestion @akash1810 - I've used this as the basis for https://github.com/guardian/cdk/pull/186/commits/dc3972c3d7f68386314c7e1644137cbc5e5f91de. To be honest, I'm not sure whether I like this more than my original implementation, but I've spent quite a bit of time experimenting and I can't find a cleaner way. Happy to revert that commit if people are not keen on this approach!